editor/src/components/editor/actions/actions.tsx

```diff
@@ -2219,78 +2222,91 @@ export const UPDATE_FNS = {
-          if (action.target == null) {
-            return {
-              components: startingComponents,
-              originalElement: null,
-              imports: success.imports,
-              insertionIndex: null,
-            }
-          } else if (insertionBehaviour.type === 'insert-as-child') {
-            return {
-              components: startingComponents,
-              imports: success.imports,
-              insertionIndex: insertionBehaviour.indexPosition ?? null,
-            }
-          } else {
-            const indexInParent = getIndexInParent(
-              success.topLevelElements,
-              EP.dynamicPathToStaticPath(action.target),
-            )
+        const updatedImports = mergeImports(
+          underlyingFilePath,
+          success.imports,
+          action.importsToAdd,
+        )
 
-            const originalElement = findJSXElementAtPath(action.target, startingComponents)
+        const { imports, duplicateNameMapping } = updatedImports
 
-            const withTargetDeleted = removeElementAtPath(
-              action.target,
-              startingComponents,
-              success.imports,
-            )
+        const fixedElement = renameJsxElementChild(action.jsxElement, duplicateNameMapping)
 
-            return {
-              components: withTargetDeleted.components,
-              originalElement: originalElement,
-              imports: withTargetDeleted.imports,
-              insertionIndex: indexInParent >= 0 ? absolute(indexInParent) : null,
-            }
-          }
+        const withInsertedElement = insertJSXElementChildren(
+          childInsertionPath(parentPath),
+          [fixedElement],
+          components,
+          action.indexPosition,
+        )
+
+        const uid = getUtopiaID(fixedElement)
+        const newPath = EP.appendToPath(parentPath, uid)
+        newSelectedViews.push(newPath)
+
+        const updatedTopLevelElements = applyUtopiaJSXComponentsChanges(
+          success.topLevelElements,
+          withInsertedElement.components,
+        )
+
+        return {
+          ...success,
+          topLevelElements: updatedTopLevelElements,
+          imports: imports,
         }
+      },
+    )
 
-        const {
-          insertionIndex,
-          originalElement,
-          components,
-          imports: workingImports,
-        } = removeElementAndReturnIndex()
+    return {
+      ...withNewElement,
+      leftMenu: { visible: editor.leftMenu.visible, selectedTab: LeftMenuTab.Navigator },
+      selectedViews: newSelectedViews,
+    }
+  },
+  REPLACE_JSX_ELEMENT: (action: ReplaceJSXElement, editor: EditorModel): EditorModel => {
+    let newSelectedViews: ElementPath[] = []
 
-        const updatedImports = mergeImports(underlyingFilePath, workingImports, action.importsToAdd)
+    const withNewElement = modifyUnderlyingParseSuccessOnly(
+      action.target,
+      editor,
+      (success, _, underlyingFilePath) => {
```

Comment:
maybe we don't even need to pass in the staticElementPath here either, since it's mostly going to be just ignored?

Reply:
good idea, removed it in https://github.com/concrete-utopia/utopia/pull/5681/commits/637888eb208208f8dab36e781cac3774c0d17179